gio/src/task.rs

```diff
@@ -382,6 +384,64 @@ impl<V: ValueType + Send> Task<V> {
 unsafe impl<V: ValueType + Send> Send for Task<V> {}
 unsafe impl<V: ValueType + Send> Sync for Task<V> {}
 
+#[derive(Debug)]
+pub struct JoinHandle<T: Send + 'static> {
+    rx: oneshot::Receiver<T>,
+    cancellable: Option<Cancellable>,
+}
+
+impl<T: Send + 'static> JoinHandle<T> {
+    #[inline]
+    fn new(cancellable: Option<Cancellable>) -> (Self, oneshot::Sender<T>) {
+        let (tx, rx) = oneshot::channel();
+        (Self { rx, cancellable }, tx)
+    }
+    #[inline]
+    pub fn cancellable(&self) -> Option<&Cancellable> {
+        self.cancellable.as_ref()
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum JoinError {
+    #[error("task cancelled")]
+    Cancelled,
+}
+
+impl<T: Send + 'static> Future for JoinHandle<T> {
+    type Output = Result<T, JoinError>;
+    #[inline]
+    fn poll(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        std::pin::Pin::new(&mut self.rx)
+            .poll(cx)
+            .map(|r| r.map_err(|_| JoinError::Cancelled))
+    }
+}
+
+pub fn spawn_blocking<T, F, P>(cancellable: Option<&P>, func: F) -> JoinHandle<T>
```

Comment:
Maybe there can be two functions to make this a bit easier. `spawn_blocking()` without cancellable and `spawn_blocking_with_cancellable()` that would be this one.

Reply:
If we keep them both, can we implement one in terms of the other so that we are sure they always behave the same?

With that said, `spawn_blocking(CancellableFuture::new(f, c))` seems as convenient as `spawn_blocking_with_cancellable(f, c)`...

Reply:
Ignore my comments, I was clearly not thinking yesterday (in my defense it was holiday :) ).

The function takes a closure, not a future. I am ok with either the proposed version or with the two variants.

Reply:
I removed the cancellable. If you need cancellation, just pass one into the closure or wrap the `JoinHandle` in an abortable future.